dev-qt/qtwebengine: Port chromium patches, fix hardened build.

lluixhi opened the pull request:
Interestingly, 5.6.0 (based on an older version of chromium) doesn't seem to need the icu patch or the madv_free patch.

The hardened patch is in a different commit because I'm not entirely sure if this is the approach we should be using.

kensington commented:
@gentoo/qt

Pesa commented on dev-qt/qtwebengine/qtwebengine-5.7.0.ebuild:
```diff
@@ -70,6 +70,11 @@ DEPEND="${RDEPEND}
 
+PATCHES=(
+	"${FILESDIR}/${PN}-5.7.0-icu58.patch"
+	"${FILESDIR}/${PN}-5.7.0-unset-madv_free.patch"
```
this patch is basically this upstream commit: http://code.qt.io/cgit/qt/qtwebengine-chromium.git/commit/?h=53-based&id=fa8cdb3a32c377b6290d0a92d2522186bcd48293

Would you mind backporting the offical commit or at least mention the commit SHA somewhere, so that it's easier to track in the future?

lluixhi replied:
Yep!

lluixhi replied:
Alright.

Pesa commented on dev-qt/qtwebengine/qtwebengine-5.6.2.ebuild:
```diff
@@ -12,7 +12,7 @@ if [[ ${QT5_BUILD_TYPE} == release ]]; then
 fi
 
-IUSE="bindist geolocation +system-ffmpeg +system-icu widgets"
+IUSE="bindist geolocation hardened +system-ffmpeg +system-icu widgets"
```
I think the correct USE flag is `pax_kernel`

kensington commented:
I asked the hardened guys to review that commit, and they said the USE flag should be called ``pax_kernel`` instead.

Pesa commented on dev-qt/qtwebengine/files/qtwebengine-paxmark-mksnapshot.patch:
```diff
@@ -0,0 +1,46 @@
++            '<(mksnapshot_exec)',
++          ],
++          'action': [
++            'bash',
```
why `bash` instead of e.g. `sh`?

lluixhi replied:
I pulled this patch off of the bug report, and the writer used bash. sh makes more sense because we aren't using any bash specific features. I'll change it.

lluixhi commented on dev-qt/qtwebengine/qtwebengine-5.6.2.ebuild:
```diff
@@ -58,6 +58,7 @@ RDEPEND="
 	x11-libs/libXScrnSaver
 	x11-libs/libXtst
 	geolocation? ( ~dev-qt/qtpositioning-${PV} )
+	pax_kernel? ( sys-apps/elfix )
```
Yeah.

lluixhi commented on dev-qt/qtwebengine/qtwebengine-5.7.0.ebuild:
```diff
@@ -70,6 +70,11 @@ DEPEND="${RDEPEND}
 "
 
+PATCHES=(
+	"${FILESDIR}/${PN}-5.7.0-icu58.patch"
```
No.

Pesa commented on dev-qt/qtwebengine/qtwebengine-5.7.0.ebuild:
```diff
@@ -68,14 +68,17 @@ DEPEND="${RDEPEND}
 PATCHES=(
 	"${FILESDIR}/${PN}-5.7.0-icu58.patch"
-	"${FILESDIR}/${PN}-5.7.0-unset-madv_free.patch"
+	"${FILESDIR}/${PN}-5.7.0-undef-madv_free.patch"
```
do this in the previous commit, i.e. when you introduce the patch in the first place

lluixhi replied:
Whoops

Pesa commented on dev-qt/qtwebengine/qtwebengine-5.7.0.ebuild:
```diff
@@ -95,3 +98,9 @@ src_configure() {
+src_install() {
+	qt5-build_src_install
+
+	pax-mark m "${ED}"/usr/$(get_libdir)/qt5/libexec/QtWebEngineProcess
```
use `"${D%/}${QT5_LIBEXECDIR}"/QtWebEngineProcess`

Also, we pax-mark unconditionally here, but we condition on `use pax_kernel` for the patch above... what's the recommended approach? The patch is needed to pax-mark a binary used during the build process, I see there's also `host-is-pax` in the eclass for this purpose, but I'm a little confused. @blueness can you shed some light please?

lluixhi replied:
About pax-marking unconditionally -- it looks like if pax-utils does not find a binary to paxmark executables, it'll simply log that the binary may not work on pax systems, so it doesn't require a system with elfix in order to function. On the other hand, the patch expects the system to have elfix, so I believe it should be masked behind the pax_kernel use flag. 

We could also get rid of the use flag and unconditionally depend on elfix for this package, but that doesn't really make sense to me.

lluixhi commented:
@Pesa I'm also noticing a bug with ffmpeg-3.x and qtwebengine-5.7.0 (I didn't notice it with 5.6.2) where qtwebengine will crash when trying to view a video.

I didn't see any commits that appeared relevant on the qtwebengine-5.6.2 branch though -- I'll try and look harder.

lluixhi commented:
Alright, I added a patch to fix ffmpeg -- it looks like the qt guys forgot to hide a stray define behind `!defined(USE_SYSTEM_FFMPEG)`.

I also added a patch to enable `-fno-delete-null-pointer-checks` from upstream. I think that's everything.

lluixhi commented:
Bump.

a17r commented:
Btw, patches still apply fine to 5.7.1 branch except the one for gcc6 (presumably it went upstream).

Pesa commented:
> Btw, patches still apply fine to 5.7.1 branch except the one for gcc6 (presumably it went upstream).

Thanks, that is what I wanted to check too, since 5.7.1 is expected to be released any day now.

lluixhi commented:
@Pesa @a17r 
That's interesting. The undef_madv_free and the gcc6 patch both come from upstream -- qtwebengine 5.7.x uses the `49-based` branch of qtwebengine-chromium.

So I believe all we'll need is the system-ffmpeg, icu58 and paxmark patches for 5.7.1

Pesa commented:
I'll try and merge this tomorrow. Sorry for the delay.

kensington commented:
@pesa I can take care of merging if you're busy, just let me know

Pesa commented:
yeah sure, go ahead, thanks.

Also, the libsrtp and pax commits should be forward-ported to the live ebuilds in the qt overlay, I guess... not sure about the patches...

kensington commented:
Thanks a lot! Picked as 3ee4ff771e9d079f4948e1190f3a9eb303621498...79d9e7f90a0e12ed552152fa6117d5df4d6b6825.
